jcl/src/java.base/share/classes/java/lang/Thread.java

```diff
@@ -1415,8 +1415,8 @@ void blockedOn(sun.nio.ch.Interruptible interruptible) {
 	int count = systemThreadGroup.activeCount() + 20;
 	Thread[] threads = new Thread[count];
 	count = systemThreadGroup.enumerate(threads);
-	java.util.Map result = new java.util.HashMap(count*4/3);
-	for (int i=0; i<count; i++) {
+	Map<Thread, StackTraceElement[]> result = new java.util.HashMap<>(count);
```

Comment:
I believe this removed initial optimization.

As per javadoc [1], _as a general rule, the default load factor (.75) offers a good tradeoff between time and space costs. Higher values decrease the space overhead but increase the lookup cost (reflected in most of the operations of the HashMap class, including get and put). The expected number of entries in the map and its load factor should be taken into account when setting its initial capacity, so as to minimize the number of rehash operations. If the initial capacity is greater than the maximum number of entries divided by the load factor, no rehash operations will ever occur._

`count*4/3` ensures no expensive rehash operations.

[1] https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/util/HashMap.html#%3Cinit%3E(int)

Reply:
The default load factor of `0.75` means that it will use an array of length `N * 1.0 / 0.75` which is (ignoring rounding and overflow) the same as `N * 4 / 3`. Providing an initial capacity 4/3 times as large as necessary would be equivalent to specifying a load factor of `0.75*0.75 = 0.625`. The original 'optimization' seems just to have been a misinterpretation of the API.

Reply:
Attached [TestHashMap.zip](https://github.com/eclipse-openj9/openj9/files/7843395/TestHashMap.zip) shows otherwise.
The test takes an integer for additional thread number, `true|false` to apply the optimization `count*4/3` or not.
The main code snippet is as following:
```
		System.out.println("actualCount = " + actualCount);
		int hashmapCount = actualCount;
		if (optimizeFlag) {
			hashmapCount = (actualCount * 4) / 3;
			System.out.println("optimizeFlag = " + optimizeFlag + ", hashmapCount((actualCount*4)/3) = " + hashmapCount);
		} else {
			System.out.println("optimizeFlag = " + optimizeFlag + ", hashmapCount(actualCount) = " + hashmapCount);
		}
		System.out.println("before new HashMap<>(" + hashmapCount + ")");
		Map<Thread, StackTraceElement[]> result = new HashMap<>(hashmapCount);
		for (int i=0; i<count; i++) {
			result.put(threads[i], threads[i].getStackTrace());
		}
		System.out.println("After new HashMap<>(" + hashmapCount + ")");
```

`java -Xtrace:print=mt{entry},methods=java/util/HashMap.resize -jar TestHashMap.zip 5 false` has following output:
```
actualCount = 7
optimizeFlag = false, hashmapCount(actualCount) = 7
before new HashMap<>(7)
02:56:09.209 0x169e6400              mt.1        > java/util/HashMap.resize()[Ljava/util/HashMap$Node; compiled method, this = 0x7fffd6240
02:56:09.210 0x169e6400              mt.1        > java/util/HashMap.resize()[Ljava/util/HashMap$Node; compiled method, this = 0x7fffd6240
After new HashMap<>(7)
```
Note: there are two entries to `HashMap.resize()` in which first one is to `Initialize` and the second one is to `double table size` [1].

When run w/ `java -Xtrace:print=mt{entry},methods=java/util/HashMap.resize -jar TestHashMap.zip 5 true`, there is only one entry to `HashMap.resize()`.
```
actualCount = 7
optimizeFlag = true, hashmapCount((actualCount*4)/3) = 9
before new HashMap<>(9)
02:59:27.445 0x107fa400              mt.1        > java/util/HashMap.resize()[Ljava/util/HashMap$Node; compiled method, this = 0x7fffd6248
After new HashMap<>(9)
```

I also added printf into `HashMap.java` and verified that the `threshold` is doubled in the second call to `resize()` when the optimization is not applied while the optimized version has right `threshold` value at first call.

I still believe `count*4/3` was an optimization.

[1] https://github.com/ibmruntimes/openj9-openjdk-jdk17/blob/732660d622a37bdffd0d3d1e8a70ea8f6981ee69/src/java.base/share/classes/java/util/HashMap.java#L666-L675

Reply:
I would consider that a bug or deficiency in HashMap (that adding `n` entries to a HashMap with initial capacity `n` involves a second call to `resize()`).

It's unlikely to affect this because of the `+ 20` above (we won't normally be using the full capacity specified).

Reply:
> It's unlikely to affect this because of the + 20 above (we won't normally be using the full capacity specified).

https://github.com/eclipse-openj9/openj9/blob/9bed0bf30491758706f0b410c677f717e2088eb3/jcl/src/java.base/share/classes/java/lang/Thread.java#L1414-L1422

The `20` is an extra room for more Threads to be created before calling enumerate(), however it is not added into initial capacity passed into `HashMap` constructor. 
`count` returned from `ThreadGroup.enumerate()` is the actual number of threads, i.e., the number of elements to be put into the newly constructed `HashMap`.
When a `HashMap` initial capacity (w/ default `0.75` loadFactor) is set same as the number of the entries to be added, the table size will have to be increased via `resize()` after initial table creation as the testcase demonstrated https://github.com/eclipse-openj9/openj9/pull/14245#discussion_r781728457.
Setting initial capacity to `count*4/3` avoids the second `resize()`.

> I would consider that a bug or deficiency in HashMap (that adding n entries to a HashMap with initial capacity n involves a second call to resize()).

Javadoc [1] states 
_When the number of entries in the hash table exceeds the product of the load factor and the current capacity, the hash table is rehashed (that is, internal data structures are rebuilt) so that the hash table has approximately twice the number of buckets._
Apparently this is by design.

[1] https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/util/HashMap.html

Reply:
Although the javadoc doesn't appear to guarantee it, the internal table size is rounded up to a power of 2.
Often, that rounding more than compensates for the 4/3 factor previously applied to the capacity argument.
On the other hand, multiplying by 4/3 can lead to that table being significantly larger than it needs to be (pushing it to an even higher power of 2).
I still believe all these sorts of details don't belong in code _using_ HashMap.

Reply:
Alright, there is no functionality issue here, if no perf test got this, it probably doesn't matter.